src/Algebra/Graph/Algorithm/Internal.hs

```diff
@@ -1,18 +1,73 @@
 module Algebra.Graph.Algorithm.Internal where
 
-import                            Algebra.Graph
-import Data.DisjointSet           (DisjointSet)
-import qualified Data.DisjointSet as DisjointSet
+import           Algebra.Graph
+import           Control.Monad
+import           Control.Monad.ST
+import           Data.Map.Strict   (Map)
+import qualified Data.Map.Strict   as Map
+import           Data.UnionFind.ST (Point)
+import qualified Data.UnionFind.ST as UF
 
--- | Extract disjoint set of connectivity components from the graph
+-- | O(s + n log n)
+-- Extract disjoint set of connectivity components from the graph
 --
--- >>> DisjointSet.pretty $ components ((1 * 2) + (3 * 4))
--- "{{1,2},{3,4}}"
+-- **NOTE:** Doesn't work properly. Check an example.
 --
--- >>> DisjointSet.pretty $ components (((1 * 2) + (3 * 4)) + (2 * 3))
--- "{{1,2,3,4}}"
-components :: Ord a => Graph a -> DisjointSet a
-components Empty           = DisjointSet.empty
-components (Vertex x)      = DisjointSet.singleton x
-components g@(Connect _ _) = DisjointSet.singletons $ vertexSet g
-components (Overlay x y)   = components x <> components y
+-- >>> components (((1 * 2) + (3 * 4)) + (2 * 3))
+-- fromList [(2,[2,1]),(3,[3,2]),(4,[4,3])]
+components :: Ord a => Graph a -> Map a [a]
+components g = runST $ do
+  (px, g') <- mkVertexPoints g
+  componentsST g'
+  componentsFromPoints px
+
+-- | O(n log n). Not sure.
+componentsFromPoints :: Ord a => [(a, Point s a)] -> ST s (Map a [a])
+componentsFromPoints vs = do
+  list <- pointsList vs
+  return $ Map.fromListWith (<>) list
+
+-- | O(n).
+pointsList :: [(a, Point s a)] -> ST s [(a, [a])]
+pointsList = mapM $ \(x, px) -> do
+  repr <- UF.descriptor px
+  return (repr, [x])
+
+-- | O(s).
+mkVertexPoints :: Eq a => Graph a -> ST s ([(a, Point s a)], (Graph (Point s a)))
+mkVertexPoints Empty = return ([], Empty)
+mkVertexPoints (Vertex v) = do
+  point <- UF.fresh v
```

Comment:
You need to check if you already created a point for this vertex. Something like this:

```suggestion
  case Map.lookup v points of
    Nothing -> do
      point <- UF.fresh v
      modify (Map.insert v point)
      return (Vertex Point)
```

So it makes sense to do it as `State (Map a (Point s a)) (Graph (Point s a))` and then to convert `Map` back into a list of pairs.

For `Ord a`, `Map` is good, but perhaps for integers from a relatively small range it is worth making a version that uses `Vector (Point s a)` instead.